lsp-ui-imenu.el

```diff
@@ -242,44 +261,47 @@ Return the updated COLOR-INDEX."
-  (let ((imenu-buffer (get-buffer-create lsp-ui-imenu-buffer-name))
-        (list imenu--index-alist))
-    (with-current-buffer imenu-buffer
-      (let* ((padding (lsp-ui-imenu--get-padding list))
-             (grouped-by-subs (-partition-by 'imenu--subalist-p list))
-             (color-index 0)
-             (bars (make-bool-vector lsp-ui-imenu--max-bars t))
-             (inhibit-read-only t))
-        (remove-overlays)
-        (erase-buffer)
-        (dolist (group grouped-by-subs)
-          (if (imenu--subalist-p (car group))
-              (setq color-index (lsp-ui-imenu--insert-items "" group padding bars 0 color-index))
-            (lsp-ui-imenu--put-separator)
-            (lsp-ui-imenu--insert-items "" group padding bars 1 color-index)
-            (setq color-index (1+ color-index))))
-        (lsp-ui-imenu-mode)
-        (when lsp-ui-imenu--custom-mode-line-format
-          (setq mode-line-format lsp-ui-imenu--custom-mode-line-format))
-        (goto-char (point-min))
-        (add-hook 'post-command-hook 'lsp-ui-imenu--post-command nil t)))
-    (let ((win (display-buffer-in-side-window imenu-buffer '((side . right)))))
-      (set-window-margins win 1)
-      (select-window win)
-      (set-window-start win 1)
-      (lsp-ui-imenu--move-to-name-beginning)
-      (set-window-dedicated-p win t)
-      ;; when `lsp-ui-imenu-window-width' is 0, fit window to buffer
-      (if (= lsp-ui-imenu-window-width 0)
-          (let ((fit-window-to-buffer-horizontally 'only))
-            (fit-window-to-buffer win)
-            (window-resize win 3 t))
-        (let ((x (- lsp-ui-imenu-window-width (window-width))))
-          (window-resize (selected-window) x t))))))
+  (lsp-ui-imenu-buffer-mode 1)
+  (let ((imenu-auto-rescan lsp-ui-imenu-auto-refresh))
+    (setq lsp-ui-imenu--origin (current-buffer))
+    (imenu--make-index-alist)
+    (let ((imenu-buffer (get-buffer-create lsp-ui-imenu-buffer-name))
+          (list imenu--index-alist))
+      (with-current-buffer imenu-buffer
+        (let* ((padding (lsp-ui-imenu--get-padding list))
+               (grouped-by-subs (-partition-by 'imenu--subalist-p list))
+               (color-index 0)
+               (bars (make-bool-vector lsp-ui-imenu--max-bars t))
+               (inhibit-read-only t))
+          (remove-overlays)
+          (erase-buffer)
+          (dolist (group grouped-by-subs)
+            (if (imenu--subalist-p (car group))
+                (setq color-index (lsp-ui-imenu--insert-items "" group padding bars 0 color-index))
+              (lsp-ui-imenu--put-separator)
+              (lsp-ui-imenu--insert-items "" group padding bars 1 color-index)
+              (setq color-index (1+ color-index))))
+          (lsp-ui-imenu-mode)
+          (when lsp-ui-imenu--custom-mode-line-format
+            (setq mode-line-format lsp-ui-imenu--custom-mode-line-format))
+          (goto-char (point-min))
+          (add-hook 'post-command-hook 'lsp-ui-imenu--post-command nil t)))
```

Comment:
Only this part of code should be called by `lsp-ui-imenu--refresh`.
See the comment in `lsp-ui-imenu--refresh`

Reply:
You meant the code from L265 to L288, right? Aren't the code 99% of the function `lsp-ui-imenu`? Any objection to just call the function? 😕

Reply:
Okay, I may know why. `lsp-ui-menu` created the window and it constantly pop out.